src/cascadia/TerminalSettingsEditor/ActionsViewModel.cpp

```diff
@@ -18,383 +24,1194 @@ using namespace winrt::Windows::UI::Xaml::Data;
-                    _Settings.ActionMap().RebindKeys(args.OldKeys(), args.NewKeys());
+                    keyChordList.emplace_back(keys);
                 }
-
-                // update view model
-                auto senderVMImpl{ get_self<KeyBindingViewModel>(senderVM) };
-                senderVMImpl->CurrentKeys(args.NewKeys());
+                auto cmdVM{ make_self<CommandViewModel>(cmd, keyChordList, *this, _AvailableActionsAndNamesMap) };
+                _RegisterCmdVMEvents(cmdVM);
+                cmdVM->Initialize();
+                commandList.push_back(*cmdVM);
             }
+        }
 
-            // If the action was changed,
-            // update the settings model and view model appropriately
-            // NOTE: no need to check for "isNewAction" here. <empty_string> != <action name> already.
-            if (args.OldActionName() != args.NewActionName())
-            {
-                // convert the action's name into a view model.
-                const auto& newAction{ _AvailableActionMap.Lookup(args.NewActionName()) };
+        std::sort(begin(commandList), end(commandList), CommandViewModelComparator{});
+        _CommandList = single_threaded_observable_vector(std::move(commandList));
+    }
+
+    void ActionsViewModel::AddNewCommand()
+    {
+        const auto newCmd = Model::Command::NewUserCommand();
+        // construct a command using the first shortcut action from our list
+        const auto shortcutAction = _AvailableActionsAndNamesMap.First().Current().Key();
+        const auto args = CascadiaSettings::GetEmptyArgsForAction(shortcutAction);
+        newCmd.ActionAndArgs(Model::ActionAndArgs{ shortcutAction, args });
+        _Settings.ActionMap().AddAction(newCmd, nullptr);
+        auto cmdVM{ make_self<CommandViewModel>(newCmd, std::vector<Control::KeyChord>{}, *this, _AvailableActionsAndNamesMap) };
+        cmdVM->IsNewCommand(true);
+        _RegisterCmdVMEvents(cmdVM);
+        cmdVM->Initialize();
+        _CommandList.Append(*cmdVM);
+        CurrentCommand(*cmdVM);
+        CurrentPage(ActionsSubPage::Edit);
+    }
+
+    void ActionsViewModel::CurrentCommand(const Editor::CommandViewModel& newCommand)
+    {
+        _CurrentCommand = newCommand;
+    }
+
+    Editor::CommandViewModel ActionsViewModel::CurrentCommand()
+    {
+        return _CurrentCommand;
+    }
 
-                // update settings model
-                _Settings.ActionMap().RegisterKeyBinding(args.NewKeys(), newAction);
+    void ActionsViewModel::CmdListItemClicked(const IInspectable& /*sender*/, const winrt::Windows::UI::Xaml::Controls::ItemClickEventArgs& e)
+    {
+        if (const auto item = e.ClickedItem())
+        {
+            CurrentCommand(item.try_as<Editor::CommandViewModel>());
+            CurrentPage(ActionsSubPage::Edit);
+        }
+    }
+
+    void ActionsViewModel::AttemptDeleteKeyChord(const Control::KeyChord& keys)
+    {
+        // Update the settings model

```

Comment:
```suggestion
        // Update the settings model
        assert(keys);
```
Thoughts on adding an `assert` here? If somebody is calling this with an empty `keys`, they're probably holding it wrong, right? So maybe we should notify the developer that they're not using it right.